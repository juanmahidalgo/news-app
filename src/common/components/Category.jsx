import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import PropTypes from 'prop-types';
import { Chip } from '@material-ui/core';

const useStyles = makeStyles(theme => ({
  chip: {
    margin: theme.spacing(0.5),
    textTransform: 'capitalize',
  },
}));

const Category = ({ category, handleClick }) => {
  const classes = useStyles();

  return (
    <Chip
        label={category}
        className={classes.chip}
        clickable
        onClick={ handleClick }
      />
  );
};

Category.propTypes = {
  category: PropTypes.string.isRequired,
  handleClick: PropTypes.func.isRequired,
};

export default Category;

