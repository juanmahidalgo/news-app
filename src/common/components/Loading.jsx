import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import CircularProgress from '@material-ui/core/CircularProgress';

const useStyles = makeStyles(theme => ({
  progress: {
    margin: theme.spacing(2),
  },
  loadingWrapper: {
    alignItems: 'center',
    display: 'flex',
    height: '100%',
    justifyContent: ' center',
  }
}));

const Loading = () => {
  const classes = useStyles();

  return (
    <div className={classes.loadingWrapper}>
      <CircularProgress className={classes.progress} />
    </div>
  );
}


export default Loading;

