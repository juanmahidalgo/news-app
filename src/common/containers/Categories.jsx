import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router';
import { compose } from 'redux';
import { makeStyles } from '@material-ui/core/styles';

import { changePage } from '../actions/news';
import { getCategories } from '../actions/categories';
import Category from '../components/Category';
import Loading from '../components/Loading';

const useStyles = makeStyles(theme => ({
  categoryWrapper: {
    margin: '2rem',
    display: 'flex',
    flexDirection: 'column',
    [theme.breakpoints.down('xs')]: {
      flexWrap: 'wrap',
      flexDirection: 'row', 
    },
    
  }
}));

export const Categories = ({ history }) => {
  const dispatch = useDispatch();
  const classes = useStyles();
  const { categories, isFetching } = useSelector(state => state.categories);
  const handleClick = category => {
    history.push(`/news/category/${category}`);
    dispatch(changePage());
  }

  useEffect(() => {
    dispatch(getCategories());
  }, []);

  if (isFetching) {
    return <Loading />;
  }

  return (
    <div className={classes.categoryWrapper}>
      {categories &&
        categories.map(
          (category, index) => (
            <Category key={index} category={category} handleClick={() => handleClick(category)} />
          )
        )
      }
    </div>
  );
}

Categories.propTypes = {
  categories: PropTypes.array,
  history: PropTypes.object.isRequired,
};

export default compose(
  withRouter,
)(Categories);

