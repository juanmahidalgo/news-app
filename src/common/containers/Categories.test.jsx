/* global jest */

import React from 'react';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import ReactTestRenderer from 'react-test-renderer';
import configureMockStore from 'redux-mock-store';
import { createMount } from '@material-ui/core/test-utils';
import { Categories } from './Categories';
import Category from '../components/Category';
import Loading from '../components/Loading';

const mockedGetCategories = jest.fn();
const mockedCategories = (props) => ({
  categories: ['Sports', 'Tech'],
  isFetching: false,
  ...props,
});
const mockedHistory = {};
const mockStore = configureMockStore([thunk]);

jest.mock('../actions/categories', () => ({
  getCategories: () => mockedGetCategories
}));

describe('<Headline />', () => {
  const mount = createMount();

  it('renders correctly', () => {
    const tree = ReactTestRenderer.create(
      <Provider store={mockStore({ categories: mockedCategories() })}>
        <Categories history={mockedHistory} />
      </Provider>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render loading component', () => {
    const wrapper = mount(
      <Provider store={mockStore({ categories: mockedCategories({ isFetching: true }) })}>
        <Categories history={mockedHistory} />
      </Provider>);
    expect(wrapper.find(Loading)).toHaveLength(1);
    expect(wrapper.find(Category)).toHaveLength(0);
  });

  it('should render 2 categories', () => {
    const wrapper = mount(
      <Provider store={mockStore({ categories: mockedCategories() })}>
        <Categories history={mockedHistory} />
      </Provider>);
    expect(wrapper.find(Category)).toHaveLength(2);
  });

  it('should call getCategories once finished first render', () => {
    const renderer = ReactTestRenderer.create(
      <Provider store={mockStore({ categories: mockedCategories })}>
        <Categories history={mockedHistory} />
      </Provider>
    );
    renderer.update();
    expect(mockedGetCategories).toHaveBeenCalled();
  });
});
