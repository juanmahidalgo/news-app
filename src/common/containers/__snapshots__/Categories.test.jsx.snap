// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Headline /> renders correctly 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "makeStyles-categoryWrapper-1",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "MuiChip-root makeStyles-chip-2 MuiChip-clickable",
          "role": "button",
          "tabindex": "0",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "MuiChip-label",
            },
            "children": Array [
              Object {
                "data": "Sports",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "span",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "MuiChip-root makeStyles-chip-2 MuiChip-clickable",
            "role": "button",
            "tabindex": "0",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "MuiChip-label",
              },
              "children": Array [
                Object {
                  "data": "Tech",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "span",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
            "role": undefined,
            "tabindex": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
            "role": undefined,
            "tabindex": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "role": undefined,
          "tabindex": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "role": undefined,
          "tabindex": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "MuiChip-root makeStyles-chip-2 MuiChip-clickable",
          "role": "button",
          "tabindex": "0",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "MuiChip-label",
            },
            "children": Array [
              Object {
                "data": "Tech",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "span",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "MuiChip-root makeStyles-chip-2 MuiChip-clickable",
            "role": "button",
            "tabindex": "0",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "MuiChip-label",
              },
              "children": Array [
                Object {
                  "data": "Sports",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "span",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
            "role": undefined,
            "tabindex": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
            "role": undefined,
            "tabindex": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "role": undefined,
          "tabindex": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "role": undefined,
          "tabindex": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
